### PR TITLE
Part a

### PR DESCRIPTION
**Pull request by sobel121**

### CODE REVIEW

File: index.html [hunk 1]
```diff
@@ -1,8 +1,8 @@
 <!DOCTYPE html>
 <html>
 <head lang="en">
-  <title>Hello React</title>
+  <title>Type Racer</title>
```
Comment (amityosef) [inline index.html:4]: fix

File: src/TargetSentence.tsx [hunk 2]
```diff
@@ -0,0 +1,12 @@
+import React from "react";
+
+export default function TargetSentence({done, todo, currentWordDone, currentWordTodo}: {done: string[], todo: string[], currentWordDone: string, currentWordTodo: string}) {
```
Comment (amityosef) [inline src/TargetSentence.tsx]: תוציא את הטייפ למשתנה

File: src/TargetSentence.tsx [hunk 3]
```diff
@@ -0,0 +1,12 @@
...
+
+export default function TargetSentence({done, todo, currentWordDone, currentWordTodo}: {done: string[], todo: string[], currentWordDone: string, currentWordTodo: string}) {
+    return (
+        <span id="targetSentence">
+            <span id="doneWords">{done.join(" ")}</span>
+            <span id="doneLettersInCurrentWord">{currentWordDone}</span>
+            <span id="todoLettersInCurrentWord">{currentWordTodo}</span>
+            <span id="todoWords">{todo.join(" ")}</span>
+        </span>
+    );
```
Comment (amityosef) [inline src/TargetSentence.tsx]: תשנה את זה לmui

File: src/appContent.css [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+#interactiveElements {
+    display: inline-flex;
+    margin-top: 10px;
+}
+
+#textInputArea {
+    width: 800px;
+    height: 32px;
+    font-size: 24px;
+}
```
Comment (amityosef) [inline src/appContent.css]: גם את זה

File: src/appContent.tsx [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+export default function AppContent() {
+    const sentenceWords = useRef<string[]>(getRandomSentenceWords());
+    const [currentTargetWordIndex, setCurrentTargetWordIndex] = useState(0);
+    const [done, setDone] = useState<string[]>([]);
+    const [todo, setTodo] = useState<string[]>(sentenceWords.current.slice(1));
+    const [currentWordDone, setCurrentWordDone] = useState("");
+    const [currentWordTodo, setCurrentWordTodo] = useState(sentenceWords.current[0]);
+
+    const handleInput = (event: React.ChangeEvent<HTMLInputElement>) => {
+        if (sentenceWords.current[currentTargetWordIndex].includes(event.currentTarget.value)) {
```
Comment (amityosef) [inline src/appContent.tsx]: destruct את המשתנים מתוך event.currentTarget.

File: src/appContent.tsx [hunk 6]
```diff
@@ -0,0 +1,59 @@
...
+    };
+
+    const restartGame = () => {
+        sentenceWords.current = getRandomSentenceWords();           
+        setTodo(sentenceWords.current.slice(1));
+        setCurrentWordTodo(sentenceWords.current[0]);
+        setCurrentWordDone(" ");
+        setDone([]);
+        setCurrentTargetWordIndex(0);
+        const typeTextArea = document.getElementById("textInputArea")! as HTMLInputElement;
```
Comment (amityosef) [inline src/appContent.tsx]: fix !

Comment (amityosef) [reply src/appContent.tsx]: document.getElementById("textInputArea") not in react

Comment (amityosef) [inline src/appContent.tsx]: as HTMLInputElement ---- dont write this

File: src/constants.ts [hunk 7]
```diff
@@ -0,0 +1,240 @@
+export const sentencesBank = [
+    "All humans are mortal beings who eventually experience the end of life.",
```
Comment (amityosef) [inline src/constants.ts]: fix

File: src/constants.ts [hunk 8]
```diff
@@ -0,0 +1,240 @@
...
+    "Vegetables and animals belong to separate kingdoms in biological classification.",
+    "Fish, as aquatic vertebrates, live in water bodies such as oceans, rivers, and lakes.",
+    "During nighttime, the absence of sunlight creates a dark environment in the outdoor surroundings.",
+    "Whole numbers, including positive integers and zero, comprise a subset of the real number system.",
+    "Snakes lack limbs, making them classified as legless reptiles with elongated bodies.",
+    "Combining blue and yellow pigments results in the creation of the color green.",
+    "The sum of angles in any triangle always equals 180 degrees, a fundamental geometric principle.",
+    "Objects near the Earth's surface experience gravitational attraction, leading to downward motion.",
+    "Insects, characterized by their exoskeleton and segmented bodies, do not possess internal skeletons.",
+    "Dogs, members of the Canidae family, belong to the class of mammals with diverse breeds.",
```
Comment (amityosef) [inline src/constants.ts]: fix

File: src/constants.ts [hunk 9]
```diff
@@ -0,0 +1,240 @@
...
+    "The Dutch Grand Prix's return to Zandvoort Circuit reignites the passion of Dutch fans, creating a sea of orange in the grandstands.",
+    "F1 drivers engage in media interviews and press conferences, providing insights into their performance, strategy, and personal experiences.",
+    "The Australian Grand Prix's Melbourne Park circuit combines a mix of fast and technical sections, offering a challenging race for drivers.",
+    "F1 safety cars play a crucial role in managing on-track incidents, ensuring the safety of drivers and marshals during races.",
+    "The Portuguese Grand Prix returned to the F1 calendar in 2020, with Portimao Circuit providing a thrilling challenge for the drivers.",
+    "F1 fans engage in spirited debates and discussions, analyzing race performances, driver rivalries, and team strategies.",
+    "nüll nüll nüll nüll nüll nüll nüll nüll nüll nüll nüll nüll nüll nüll nüll nüll nüll nüll nüll nüll nüll nüll nüll nüll nüll nüll.",
+    "fãilure fãilure fãilure fãilure fãilure fãilure fãilure fãilure fãilure fãilure fãilure fãilure fãilure fãilure fãilure fãilure fãilure."
```
Comment (amityosef) [inline src/constants.ts]: בעיברית בבקשה

**Review by amityosef — CHANGES_REQUESTED** [review]

File: src/App.tsx [hunk 10]
```diff
@@ -1,7 +1,12 @@
-import React from 'react'
+import React from "react";
+require("./App.css");
```
Comment (itayboop) [inline src/App.tsx]: require??? are you in js?
`import 'App.css'`

File: src/appContent.tsx [hunk 11]
```diff
@@ -0,0 +1,62 @@
+import React, { useRef, useState } from "react";
+import { getRandomSentenceWords, isWordComplete } from "./utils";
+import TargetSentence from "./TargetSentence";
+require('./appContent.css')
```
Comment (itayboop) [inline src/appContent.tsx]: again
import

File: src/appContent.tsx [hunk 12]
```diff
@@ -0,0 +1,62 @@
...
+
+export default function AppContent() {
+    const sentenceWords = useRef<string[]>(getRandomSentenceWords());
+    const textArea = useRef<HTMLInputElement>(null);
+    const [currentTargetWordIndex, setCurrentTargetWordIndex] = useState(0);
+    const [done, setDone] = useState<string[]>([]);
+    const [todo, setTodo] = useState<string[]>(sentenceWords.current.slice(1));
+    const [currentWordDone, setCurrentWordDone] = useState("");
+    const [currentWordTodo, setCurrentWordTodo] = useState(sentenceWords.current[0]);
```
Comment (itayboop) [inline src/appContent.tsx]: you dont need this state

File: index.html (hunk 1 above)
Comment (10612) [inline index.html:4]: This should be implemented in a component

File: src/App.tsx (hunk 10 above)
Comment (10612) [inline src/App.tsx]: Why not `import "./App.css"`?

File: src/utils.ts [hunk 13]
```diff
@@ -0,0 +1,8 @@
+import { sentencesBank } from "./constants";
+
+export const getRandomSentenceWords = () => sentencesBank[Math.floor(Math.random() * sentencesBank.length)].split(" ");
+
+export const isWordComplete = (value: string, sentenceWords: string[], currentTargetWordIndex: number) => {
+    return value === sentenceWords[currentTargetWordIndex] + " " 
+    || (value ===sentenceWords[currentTargetWordIndex] && sentenceWords.length === currentTargetWordIndex + 1)
```
Comment (itayboop) [inline src/utils.ts]: split those into several lines so it will be more readable

**Review by itayboop — CHANGES_REQUESTED** [review]

File: src/RestartGameButton.tsx
Comment (10612) [inline src/RestartGameButton.tsx:1]: Delete empty file

File: src/constants.ts
Comment (10612) [inline src/constants.ts:1]: Give a more indicative name

File: src/utils.ts [hunk 14]
```diff
@@ -0,0 +1,8 @@
+import { sentencesBank } from "./constants";
+
+export const getRandomSentenceWords = () => sentencesBank[Math.floor(Math.random() * sentencesBank.length)].split(" ");
+
+export const isWordComplete = (value: string, sentenceWords: string[], currentTargetWordIndex: number) => {
```
Comment (10612) [inline src/utils.ts]: Change value to a more indicative name

File: src/TargetSentence.tsx [hunk 15]
```diff
@@ -0,0 +1,14 @@
+import React from "react";
+
+interface TargetSentenceProps {done: string[], todo: string[], currentWordDone: string, currentWordTodo: string};
```
Comment (10612) [inline src/TargetSentence.tsx]: {
    prop1: type,
    prop2: type,
    ...
}

File: src/appContent.tsx [hunk 16]
```diff
@@ -0,0 +1,62 @@
...
+            textArea.current.focus();
+        }
+    };
+
+    return (
+        <>
+            <TargetSentence todo={todo} done={done} currentWordDone={currentWordDone} currentWordTodo={currentWordTodo} />
+            <span id="intearctiveElements">
+                <input type="textArea" onChange={handleInput} ref={textArea} id="textInputArea"></input>
+                <button onClick={restartGame} id="restartGameButton">restart game</button>
```
Comment (10612) [inline src/appContent.tsx]: Put "restart game" in a `strings.ts` file

File: src/App.tsx [hunk 17]
```diff
@@ -1,7 +1,12 @@
...
+import React from "react";
+import "./App.css";
+import AppContent from "./appContent";
 
 export default function App() {
     return (
-        <h1>hi</h1>
-    )
+        <div id="appContainer">
+            <h1>type racer</h1>
```
Comment (10612) [inline src/App.tsx]: pur "type racer" in a `strings.ts` file

File: src/appContent.tsx [hunk 18]
```diff
@@ -0,0 +1,62 @@
...
+            textArea.current.value = ""; 
+            textArea.current.focus();
+        }
+    };
+
+    return (
+        <>
+            <TargetSentence todo={todo} done={done} currentWordDone={currentWordDone} currentWordTodo={currentWordTodo} />
+            <span id="intearctiveElements">
+                <input type="textArea" onChange={handleInput} ref={textArea} id="textInputArea"></input>
```
Comment (10612) [inline src/appContent.tsx]: Move input to a separate component

**Review by 10612 — CHANGES_REQUESTED** [review]
Put each component in a separate file

File: package.json [hunk 19]
```diff
@@ -23,6 +23,7 @@
     "webpack-dev-server": "^4.15.1"
   },
   "dependencies": {
+    "module-name": "^0.0.1-security",
```
Comment (odev954) [inline package.json]: what is this dep?

File: src/components/appContent/appContent.tsx [hunk 20]
```diff
@@ -0,0 +1,52 @@
...
+            textArea.current.value = "";
+            textArea.current.focus();
+        }
+    };
+
+    return (
+        <>
+            <TargetSentence
+                done={done}
+                todo={getTodoWords(done.length + 1, sentenceWords.current)}
```
Comment (odev954) [inline src/components/appContent/appContent.tsx]: use useMemo hook

File: src/components/appContent/appContent.tsx [hunk 21]
```diff
@@ -0,0 +1,52 @@
...
+                currentWordDone={currentWordDoneCharacters}
+                currentWordTodo={getCurrentWordTodoCharacters(
+                    currentWordDoneCharacters.length,
+                    sentenceWords.current[currentTargetWordIndex]
+                )}
```
Comment (odev954) [inline src/components/appContent/appContent.tsx]: >use useMemo hook

as well

File: src/components/appContent/appContent.tsx [hunk 22]
```diff
@@ -0,0 +1,52 @@
...
+        setCurrentWordDoneCharacters("");
+        setDone([]);
+        setCurrentTargetWordIndex(0);
+
+        if (textArea.current) {
+            textArea.current.disabled = false;
+            textArea.current.value = "";
+            textArea.current.focus();
+        }
+    };
```
Comment (odev954) [inline src/components/appContent/appContent.tsx]: use useCallback hook

File: src/components/targetSentence/TargetSentence.tsx [hunk 23]
```diff
@@ -0,0 +1,20 @@
...
+    done: string[],
+    todo: string[], 
+    currentWordDone: string, 
+    currentWordTodo: string
+};
+
+export default function TargetSentence({done, todo, currentWordDone, currentWordTodo}: TargetSentenceProps) {
+    return (
+        <span id="targetSentence">
+            <span className="targetSentencePart" id="doneWords">{done.join(" ") + " "}</span>
```
Comment (odev954) [inline src/components/targetSentence/TargetSentence.tsx]: put `done.join(" ") + " "` inside useMemo hook

File: src/components/targetSentence/TargetSentence.tsx [hunk 24]
```diff
@@ -0,0 +1,20 @@
...
+            <span className="targetSentencePart" id="doneLettersInCurrentWord">{currentWordDone}</span>
+            <span className="targetSentencePart" id="todoLettersInCurrentWord">{currentWordTodo}</span>
+            <span className="targetSentencePart" id="todoWords">{" " + todo.join(" ")}</span>
```
Comment (odev954) [inline src/components/targetSentence/TargetSentence.tsx]: put `" " + todo.join(" ")` inside useMemo hook

File: src/components/typeInput/TypeInput.tsx [hunk 25]
```diff
@@ -0,0 +1,40 @@
+import React from "react";
+import { isWordComplete } from "./utils";
+
+interface TypeInputProps {
+    sentenceWords: string[], 
+    currentTargetWordIndex: number, 
+    setDone: React.Dispatch<React.SetStateAction<string[]>>, 
+    setCurrentWordDoneCharacters: React.Dispatch<React.SetStateAction<string>>, 
+    setCurrentTargetWordIndex: React.Dispatch<React.SetStateAction<number>>
```
Comment (odev954) [inline src/components/typeInput/TypeInput.tsx]: you are cupling your code to the type of ` React.Dispatch`, just use regular callback syntax.
example:
```suggestion
    setDone: (done: string[]) => void,  
    setCurrentWordDoneCharacters: React.Dispatch<React.SetStateAction<string>>, //do this here too
    setCurrentTargetWordIndex: React.Dispatch<React.SetStateAction<number>> //do this here too
```

File: src/components/typeInput/TypeInput.tsx [hunk 26]
```diff
@@ -0,0 +1,40 @@
...
+
+interface TypeInputProps {
+    sentenceWords: string[], 
+    currentTargetWordIndex: number, 
+    setDone: React.Dispatch<React.SetStateAction<string[]>>, 
+    setCurrentWordDoneCharacters: React.Dispatch<React.SetStateAction<string>>, 
+    setCurrentTargetWordIndex: React.Dispatch<React.SetStateAction<number>>
+}
+
+function TypeInput({sentenceWords, currentTargetWordIndex, setDone, setCurrentWordDoneCharacters, setCurrentTargetWordIndex}:TypeInputProps, textArea: React.ForwardedRef<HTMLInputElement>) {
```
Comment (odev954) [inline src/components/typeInput/TypeInput.tsx]: why you are getting textArea in separate of props? this is not conventional

File: src/components/typeInput/TypeInput.tsx [hunk 27]
```diff
@@ -0,0 +1,40 @@
...
+
+            setCurrentTargetWordIndex((index) => index + 1);
+            event.currentTarget.value = "";
+            setCurrentWordDoneCharacters("");
+
+            if (sentenceWords.length === currentTargetWordIndex + 1) {
+                event.currentTarget.disabled = true;
+            }
+        }
+    };
```
Comment (odev954) [inline src/components/typeInput/TypeInput.tsx]: use useCallback hook

**Review by odev954 — COMMENTED** [review]
* conventions
* hooks

File: src/components/typeInput/TypeInput.tsx [hunk 28]
```diff
@@ -0,0 +1,41 @@
+import React, { useCallback } from "react";
+import { isWordComplete } from "./utils";
+
+interface TypeInputProps {
+    sentenceWords: string[], 
+    currentTargetWordIndex: number, 
+    setDone: (callback: (value: string[]) => string[]) => void;
+    setCurrentWordDoneCharacters: React.Dispatch<React.SetStateAction<string>>, 
+    setCurrentTargetWordIndex: React.Dispatch<React.SetStateAction<number>>,
```
Comment (odev954) [inline src/components/typeInput/TypeInput.tsx]: decouple

File: src/components/appContent/appContent.tsx [hunk 29]
```diff
@@ -0,0 +1,70 @@
...
```
Comment (odev954) [inline src/components/appContent/appContent.tsx]: use callback

**Review by odev954 — COMMENTED** [review]
* use callback
* decouple

**Review by odev954 — APPROVED** [review]

**Pull request merged**